tests/server.c

```diff
@@ -0,0 +1,359 @@
+#define perror(str) fprintf(stderr, str ": %d.\n", WSAGetLastError())
+#define strerror(e) ws_strerror(e)
+#ifndef STDOUT_FILENO
+#define STDOUT_FILENO 1 /* MinGW has this */
+#endif /* !STDOUT_FILENO */
+#endif /* _WIN32 */
+
+/* crustls.h is autogenerated in the Makefile using cbindgen. */
+#include "crustls.h"
+#include "common.h"
+
+enum crustls_demo_result
+read_file(const char *filename, char *buf, size_t buflen, size_t *n)
+{
+  FILE *f = fopen(filename, "r");
+  if(f == NULL) {
+    fprintf(stderr, "%s\n", strerror(errno));
+    return CRUSTLS_DEMO_ERROR;
+  }
+  *n = fread(buf, 1, buflen, f);
+  if(!feof(f)) {
+    fprintf(stderr, "%s\n", strerror(errno));
+    return CRUSTLS_DEMO_ERROR;
+  }
+  return CRUSTLS_DEMO_OK;
+}
+
+typedef enum exchange_state {
+  READING_REQUEST,
+  SENT_RESPONSE
+} exchange_state;
+
+/*
+ * Do one read from the socket, and process all resulting bytes into the
+ * rustls_connection, then copy all plaintext bytes from the session to stdout.
+ * Returns:
+ *  - CRUSTLS_DEMO_OK for success
+ *  - CRUSTLS_DEMO_AGAIN if we got an EAGAIN or EWOULDBLOCK reading from the
+ *    socket
+ *  - CRUSTLS_DEMO_EOF if we got EOF
+ *  - CRUSTLS_DEMO_ERROR for other errors.
+ */
+enum crustls_demo_result
+do_read(struct conndata_t *conn, struct rustls_connection *rconn)
+{
+  int err = 1;
+  int result = 1;
+  size_t n = 0;
+
+  err = rustls_connection_read_tls(rconn, read_cb, conn, &n);
+  if(err == EAGAIN || err == EWOULDBLOCK) {
+    fprintf(stderr,
+            "reading from socket: EAGAIN or EWOULDBLOCK: %s\n",
+            strerror(errno));
+    return CRUSTLS_DEMO_AGAIN;
+  }
+  else if(err != 0) {
+    fprintf(stderr, "reading from socket: errno %d\n", err);
+    return CRUSTLS_DEMO_ERROR;
+  }
+
+  if (n == 0) {
+    return CRUSTLS_DEMO_EOF;
+  }
+  fprintf(stderr, "read %ld bytes from socket\n", n);
+
+  result = rustls_connection_process_new_packets(rconn);
+  if(result != RUSTLS_RESULT_OK) {
+    print_error("in process_new_packets", result);
+    return CRUSTLS_DEMO_ERROR;
+  }
+
+  result = copy_plaintext_to_buffer(conn);
+  if(result != CRUSTLS_DEMO_CLOSE_NOTIFY) {
+    fprintf(stderr, "do_read returning %d\n", result);
+    return result;
+  }
+
+  char buf[2048];
+  /* If we got a close_notify, verify that the sender then
+   * closed the TCP connection. */
+  n = read(conn->fd, buf, sizeof(buf));
+  if(n != 0 && errno != EWOULDBLOCK) {
+    fprintf(stderr, "read returned %ld after receiving close_notify: %s\n", n, strerror(errno));
+    return CRUSTLS_DEMO_ERROR;
+  }
+  return CRUSTLS_DEMO_CLOSE_NOTIFY;
+}
+
+void
+send_response(struct conndata_t *conn) {
+  struct rustls_connection *rconn = conn->rconn;
+  const char* response = "HTTP/1.1 200 OK\r\nContent-Length: 6\r\n\r\nhello\n";
+  size_t n;
+  rustls_connection_write(rconn, (const uint8_t *)response, strlen(response), &n);
+  if (n != strlen(response)) {
+    fprintf(stderr, "failed to write all response bytes. wrote %ld\n", n);
+    abort();
```

Comment:
It doesn't really matter since this is a test program, but `abort()` seems a little harsh since it'd be pretty easy to return an error to the single callsite in `handle_conn`, which already returns an error.